client/consensus/babe/src/lib.rs

```diff
@@ -1036,16 +1027,6 @@ where
 		create_inherent_data_providers: CIDP::InherentDataProviders,
 		execution_context: ExecutionContext,
 	) -> Result<(), Error<Block>> {
-		if let Err(e) = self.can_author_with.can_author_with(&block_id) {
```

Comment:
Shoudn't there still be runtime API compatibility check before calling `check_inherents_with_context`?

Reply:
The runtime api should be compatible and this should always have returned `Ok(())` for Polkadot.

Reply:
Why does aura verifier have this check then?
https://github.com/paritytech/substrate/blob/3449fb003534c0c361a92a2d1fe03abc80e4997d/client/consensus/aura/src/import_queue.rs#L229-L235

> The runtime api should be compatible and this should always have returned Ok(()) for Polkadot.

Except it did not when the state was not available. E.g. when importing historical blocks after the warp sync. These blocks are not executed at all and there's no parent state to call the runtime on.

Reply:
> Why does aura verifier have this check then?

I mean it is compatible or better, we at least check and skip if it is not compatible. I don't see any problem there? Especially nothing that would have been solved by `can_author_with`.



> Except it did not when the state was not available. E.g. when importing historical blocks after the warp sync. These blocks are not executed at all and there's no parent state to call the runtime on.

This sounds like the bug, however I wonder how this worked before? Because `can_author_with` should not have protected you against trying to call into a non available state?

Reply:
`can_author_with` returns an error when there's no state. And the code above simply checked for any error. I'll make a PR to skip this check if the block is imported with no execution.